### ISSUE TITLE
Deprecate encryption/decryption support in twisted.persisted.sob

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8081]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8081|
|Type|enhancement|
|Created|2015-11-03 09:12:06Z|
|Branch|https://github.com/twisted/twisted/tree/persisted-sob-AES-deprecate-8081|

twisted/persisted/sob.py depends on PyCrypto for AES and we are on the path of removing PyCrypto as a dependency.

glyph

> We need to just remove that functionality. It's not a good way to encrypt stuff. It's using MD5 as a KDF, it's doing its own wacky custom padding with spaces, it's using ECB mode... it basically isn't even encryption at all. And I hope nobody is actually using pickling of application state any more - we should deprecate persistence entirely.

<details><summary>Searchable metadata</summary>

```
trac-id__8081 8081
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__branches_persisted_sob_AES_deprecate_8081 branches/persisted-sob-AES-deprecate-8081
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1446541926680342 1446541926680342
changetime__1447341269943616 1447341269943616
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

(In [46040]) Branching to persisted-sob-AES-deprecate-8081.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

# Scope

This branch deprecates the encryption usage in twisted.persisted.sob.

The encryption is weak and we want to remote the PyCrypto dependency. 


# Changes

raise a warning when encryption/decryption private methods are used.

Move all crypto related tests into a common test suite.


# How to test

Check that changes make sense.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban removed owner|
|-|-|

I have run the test on my local system and I hope they will also pass on builders.

Please see previous comment and review.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi, thanks for this.

- The deprecation tests should flush the warnings from it specifically, as in https://github.com/twisted/twisted/blob/trunk/twisted/scripts/test/test_scripts.py#L198 -- if you don't, sometimes it will fail if a warning is raised elsewhere and wasn't flushed.

- In the docstring, "The encyption used by these method is weak." should be "The encryption used by these methods are weak."

- Rather than have Crypto and skipCrypto, elsewhere we have a try, except ImportError, else pattern where there is only skipCrypto, which is set to None if it imports fine. Then there's only one variable required.

Please fix these small issues up and merge. It's good to get this junk deprecated.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Thanks for the review.

Will also update the proposed deprecation docs to include this note.

I think that we need a better API on top of flushWarnings which will automatically filter only warnings from the test.

Please note that my tests also check for the lengths of all warnings... so it should fail it if catches other non-related warnings.

Will fix and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

flushWarnings with a target is a PITA but I have fixed it.

I agree that flushAll is bad.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46165]) Merge persisted-sob-AES-deprecate-8081: Deprecate encryption/decryption support in twisted.persisted.sob.

Author: adiroiban
Reviewer: hawkowl
Fixes: [#8081](https://github.com/twisted/twisted/issues/8081)
